doc/internals/export.rst

```diff
@@ -0,0 +1,98 @@
+.. _export:
+
+Exporting Symbols
+=================
+
+The arbor libraries are compiled with hidden visibility by default. Thus, symbols which are part of
+the public API need to be marked explicitely as visible. Arbor provides a couple of macros to
+annotate functions and classes which are defined in the header file ``export.hpp`` in each library's
+include directory, i.e. ``include/arbor/export.hpp``. These header files are generated at configure
+time based on the build variant, compiler and platform.
+
+Macro Descripiton
+-----------------
+
+.. c:macro:: ARB_LIBNAME_API
+
+    Here ``LIBNAME`` is a placeholder for the library's name: ``ARB_ARBOR_API`` for the main arbor
+    library, ``ARB_ARBORIO_API`` for arborio, etc. This macro is intended to annotate functions,
+    classes and structs which need to be accessible when using the library. Note, that it expands to
+    different values when arbor is being built vs. when arbor is being used by an application. Below
+    we list the places where the macro needs to be added or can be safely omitted (we assume all of
+    the symbols below are part of the public API).
+
+    .. code-block:: cpp
+        :caption: header.hpp
+
+        // free function declaration
+        ARB_ARBOR_API void foo();
+
+        // free function (inline)
+        void bar(int i) { /* ... */ }
+
+        // function template (inline)
+        template<typename T>
+        void baz(T i) { /* ... */ }
+
+        // class declaration
+        // note: this will make all member symbols visible
```

Comment:
member symbols will become visible as per class visibility directives or do you mean that those are overridden as well?

Reply:
I don't quite understand. All members of a class (member functions and members) will be implicitly annotated with hidden=default (i.e. visible) if the class is made visible through the macro.

Reply:
```
class B {
   int myint;
};
```
`myint` is hidden because class members are hidden by default. `B` is hidden from the public API, because it isn't annotated to be visible. Right?
```
class ARB_ARBOR_API B {
   int myint;
};
```
I would expect `myint` to be hidden, even though having annotated `B` to be visible. Is that so?